overlord/snapstate/cookies.go

```diff
@@ -0,0 +1,169 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/strutil"
+)
+
+func cookies(st *state.State) (map[string]string, error) {
+	var snapCookies map[string]string
+	if err := st.Get("snap-cookies", &snapCookies); err != nil {
+		if err != state.ErrNoState {
+			return nil, fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		snapCookies = make(map[string]string)
+	}
+	return snapCookies, nil
+}
+
+// SyncCookies creates snap cookies for snaps that are missing them (may be the case for snaps installed
+// before the feature of running snapctl outside of hooks was introduced, leading to a warning
+// from snap-confine).
+// It is the caller's responsibility to lock state before calling this function.
+func (m *SnapManager) SyncCookies(st *state.State) error {
+	var snapNames map[string]*json.RawMessage
+	if err := st.Get("snaps", &snapNames); err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	snapCookies, err := cookies(st)
+	if err != nil {
+		return err
+	}
+
+	snapsWithCookies := make(map[string]bool)
+	for _, snap := range snapCookies {
+		if _, ok := snapNames[snap]; !ok || snapsWithCookies[snap] {
```

Comment:
This checks if there is a cookie for a not-installed snap or if there are multiple cookies for the same snap, right? if, maybe a tiny comment here to make reading it easier?

Reply:
Done.